pkg/volume/cinder/attacher.go

```diff
@@ -0,0 +1,240 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cinder
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/keymutex"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+type cinderDiskAttacher struct {
+	host volume.VolumeHost
+}
+
+var _ volume.Attacher = &cinderDiskAttacher{}
+
+var _ volume.AttachableVolumePlugin = &cinderPlugin{}
+
+const (
+	checkSleepDuration = time.Second
+)
+
+// Singleton key mutex for keeping attach/detach operations for the same PD atomic
+var attachDetachMutex = keymutex.NewKeyMutex()
+
+func (plugin *cinderPlugin) NewAttacher() (volume.Attacher, error) {
+	return &cinderDiskAttacher{host: plugin.host}, nil
+}
+
+func (attacher *cinderDiskAttacher) Attach(spec *volume.Spec, hostName string) error {
+	volumeSource, _ := getVolumeSource(spec)
+	VolumeID := volumeSource.VolumeID
+
+	// Block execution until any pending detach operations for this PD have completed
+	attachDetachMutex.LockKey(VolumeID)
+	defer attachDetachMutex.UnlockKey(VolumeID)
+
+	cloud, err := getCloudProvider(attacher.host.GetCloudProvider())
+	if err != nil {
+		return err
+	}
+	instanceid, err := cloud.InstanceID()
+	if err != nil {
+		return err
+	}
+	_, err = cloud.AttachDisk(instanceid, VolumeID)
+	return err
+}
+
+func (attacher *cinderDiskAttacher) WaitForAttach(spec *volume.Spec, timeout time.Duration) (string, error) {
+	cloud, err := getCloudProvider(attacher.host.GetCloudProvider())
```

Comment:
@rootfs Please make sure to address this ASAP in a follow up PR. For 1.3 we do to not want any cloud provider dependency for WaitForAttach/Detach and Mount/Unmount.